crates/matrix-sdk-crypto/src/types/mod.rs

```diff
@@ -27,6 +27,167 @@ mod cross_signing_key;
+
+impl Signatures {
+    /// Create a new, empty, signatures collection.
+    pub fn new() -> Self {
+        Signatures(Default::default())
+    }
+
+    /// Add the given signature from the given signer and the given key_id to
+    /// the collection.
+    pub fn add_signature(
+        &mut self,
+        signer: OwnedUserId,
+        key_id: OwnedDeviceKeyId,
+        signature: Ed25519Signature,
+    ) -> Option<Signature> {
+        self.0.entry(signer).or_insert_with(Default::default).insert(key_id, signature.into())
+    }
+
+    /// Try to find an Ed25519 signature from the given signer with the given
+    /// key id.
+    pub fn get_signature(&self, signer: &UserId, key_id: &DeviceKeyId) -> Option<Ed25519Signature> {
+        self.get(signer)?.get(key_id)?.ed25519()
+    }
+
+    /// Get the map of signatures that belong to the given user.
+    pub fn get(&self, signer: &UserId) -> Option<&BTreeMap<OwnedDeviceKeyId, Signature>> {
+        self.0.get(signer)
+    }
+
+    /// Remove all the signatures we currently hold.
+    pub fn clear(&mut self) {
+        self.0.clear()
+    }
+
+    /// Do we hold any signatures or is our collection completely empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// How many signatures do we currently hold.
+    pub fn signature_count(&self) -> usize {
+        self.0.iter().map(|(_, u)| u.len()).sum()
+    }
+}
+
+impl Default for Signatures {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl IntoIterator for Signatures {
+    type Item = (OwnedUserId, BTreeMap<OwnedDeviceKeyId, Signature>);
+
+    type IntoIter =
+        std::collections::btree_map::IntoIter<OwnedUserId, BTreeMap<OwnedDeviceKeyId, Signature>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'de> Deserialize<'de> for Signatures {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let map: BTreeMap<OwnedUserId, BTreeMap<OwnedDeviceKeyId, String>> =
+            serde::Deserialize::deserialize(deserializer)?;
+
+        Ok(Signatures(
+            map.into_iter()
+                .map(|(user, signatures)| {
+                    Ok((
+                        user,
+                        signatures
+                            .into_iter()
+                            .map(|(key_id, s)| {
+                                let algorithm = key_id.algorithm();
+                                let signature =
+                                    (algorithm, s).try_into().map_err(serde::de::Error::custom)?;
+
+                                Ok((key_id, signature))
+                            })
+                            .collect::<Result<BTreeMap<_, _>, _>>()?,
+                    ))
+                })
+                .collect::<Result<_, _>>()?,
+        ))
+    }
+}
+
+impl TryFrom<(DeviceKeyAlgorithm, String)> for Signature {
```

Comment:
I'm not sure this should be part of the public interface, it seems rather ad-hoc. With my previous comment I was mostly saying we might want to nest thing a bit less in `deserialize`. I'll leave a rough suggestion.